File: src/cli/container_builder.py
"""On-demand container-based Linux builder for the macOS container runtimes.

On podman / Apple Container, when a `packages:` build isn't cached, nix must
offload to Linux.  Instead of a separate QEMU VM (``builder.py`` /
``darwin.linux-builder`` — the roadmap fallback), run a tiny nix+sshd builder
as a container **on the runtime that's already up**: ephemeral, no second
hypervisor, zero idle RAM.  The image is ``packages.builderImage`` (flake),
published to ``ghcr.io/mschulkind-oss/yolo-jail-builder`` and pulled on the Mac.

Mechanism (proven end-to-end on podman): the container runs sshd with
``ForceCommand nix-daemon --stdio``; nix's ssh-ng remote-builder protocol talks
to it, so a host ``nix build --builders 'ssh-ng://root@<addr> …'`` runs the
build inside the container and copies the result back.

Design (all podman-verifiable on Linux; AC differs only in how the host
reaches the container — see ``reachable_address``):
  * **pure** — argv/URI/`builders` builders, key-path helpers.
  * **impure** — pull / run / wait-reachable / stop; the ``builder_session``
    context manager ties them into a start→use→teardown lifecycle.

The host-side private key is generated per-workspace under GLOBAL_STORAGE and
its public half is passed to the container via ``YOLO_BUILDER_PUBKEY`` at run
time (the published image is keyless — one image serves every user).
"""

from __future__ import annotations

import contextlib
import os
import socket
import subprocess
import time
from typing import Iterator, List, Optional, Tuple

from .paths import GLOBAL_STORAGE

# The published builder image (keyless; key injected at run time).  Pinned by
# tag; a digest pin is a hardening follow-up.
BUILDER_IMAGE = "ghcr.io/mschulkind-oss/yolo-jail-builder:latest"
BUILDER_CONTAINER = "yolo-linux-builder"  # fixed name → idempotent reuse/cleanup
BUILDER_SSH_USER = "root"  # the image logs in as root (owns /nix/store)
BUILDER_GUEST_PORT = 22
# Host port we publish the container's sshd to (podman `-p`).  Fixed + high to
# avoid collisions; AC ignores this and exposes the container's own VM IP:22.
BUILDER_HOST_PORT = 31022

# Per-workspace host-daemon key: its PUBLIC half is authorized in the container.
BUILDER_KEY_DIR = GLOBAL_STORAGE / "linux-builder-container"
BUILDER_KEY = BUILDER_KEY_DIR / "id_ed25519"

BUILDER_START_TIMEOUT_S = 60
BUILDER_POLL_INTERVAL_S = 1.0


# ── pure helpers ─────────────────────────────────────────────────────────────


def ensure_keypair(_run=None) -> str:
    """Generate the host-daemon builder keypair if absent; return the PUBLIC key.

    ed25519, no passphrase, 0600 private — the private key stays on the host
    (nix ssh's with it); only the public half enters the container.  Idempotent.
    """
    run = _run or subprocess.run
    if BUILDER_KEY.exists() and BUILDER_KEY.with_suffix(".pub").exists():
        return BUILDER_KEY.with_suffix(".pub").read_text().strip()
    BUILDER_KEY_DIR.mkdir(parents=True, exist_ok=True)
    run(
        [
            "ssh-keygen",
            "-t",
            "ed25519",
            "-N",
            "",
            "-q",
            "-f",
            str(BUILDER_KEY),
            "-C",
            "yolo-linux-builder",
        ],
        check=True,
    )
    try:
        os.chmod(BUILDER_KEY, 0o600)
    except OSError:
        pass
    return BUILDER_KEY.with_suffix(".pub").read_text().strip()


def pull_argv(runtime: str, image: str = BUILDER_IMAGE) -> List[str]:
    """argv to pull the builder image on the given runtime. Pure."""
    if runtime == "container":
        return ["container", "image", "pull", image]
    return [runtime, "pull", image]


def run_argv(
    runtime: str,
    pubkey: str,
    image: str = BUILDER_IMAGE,
    name: str = BUILDER_CONTAINER,
    host_port: int = BUILDER_HOST_PORT,
) -> List[str]:
    """argv to start the builder container detached. Pure.

    podman publishes sshd to 127.0.0.1:<host_port>; Apple Container has no
    ``-p`` (each container gets its own VM IP), so we omit the publish there and
    the caller discovers the container's address via ``reachable_address``.
    """
    common = [
        "run",
        "-d",
        "--rm",
        "--name",
        name,
        "-e",
        f"YOLO_BUILDER_PUBKEY={pubkey}",
    ]
    if runtime == "container":
        return ["container", *common, image]
    # podman: publish the guest sshd to a fixed host loopback port.
    return [
        runtime,
        *common,
        "-p",
        f"127.0.0.1:{host_port}:{BUILDER_GUEST_PORT}",
        image,
    ]


def builder_uri(host: str, port: int = BUILDER_HOST_PORT) -> str:
    """The ssh-ng store/builder URI nix uses to reach the container. Pure."""
    return f"ssh-ng://{BUILDER_SSH_USER}@{host}:{port}?ssh-key={BUILDER_KEY}"


def builders_line(host: str, port: int = BUILDER_HOST_PORT, max_jobs: int = 4) -> str:
    """A nix ``--builders`` spec pointing at the container. Pure.

    Format: ``ssh-ng://user@host system key maxjobs`` — the host daemon offloads
    matching-system builds here.  System is fixed to aarch64-linux (the arch a
    Mac needs; the container image is aarch64-linux).
    """
    return f"ssh-ng://{BUILDER_SSH_USER}@{host}:{port} aarch64-linux {BUILDER_KEY} {max_jobs}"


def nix_ssh_opts() -> str:
    """NIX_SSHOPTS for talking to an ephemeral container (no host-key pinning).

    The container regenerates its host key each boot, so pinning is pointless;
    accept-new + throwaway known-hosts avoids interactive prompts and stale-key
    failures.  This is a build-only localhost/VM-IP endpoint.
    """
    return "-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"


# ── impure lifecycle ─────────────────────────────────────────────────────────


def reachable(host: str, port: int, timeout: float = 1.0) -> bool:
    """True if something accepts TCP at host:port (sshd up). Never raises."""
    try:
        with socket.create_connection((host, port), timeout=timeout):
            return True
    except OSError:
        return False


def reachable_address(
    runtime: str, name: str = BUILDER_CONTAINER
) -> Optional[Tuple[str, int]]:
    """Where the host reaches the running builder's sshd — runtime-specific.

    podman: the published loopback port (127.0.0.1:BUILDER_HOST_PORT).
    Apple Container: the container's own VM IP on :22, read from ``container
    ls`` (AC has no host port-publish; the per-container VM IP is directly
    reachable — verified on real HW at 192.168.64.2:22).
    Returns ``(host, port)`` or None if it can't be determined.
    """
    if runtime != "container":
        return "127.0.0.1", BUILDER_HOST_PORT
    try:
        r = subprocess.run(
            ["container", "ls"], capture_output=True, text=True, timeout=10
        )
    except (OSError, subprocess.SubprocessError):
        return None
    if r.returncode != 0:
        return None
    for line in r.stdout.strip().splitlines()[1:]:  # skip header
        parts = line.split()
        if parts and parts[0] == name:
            # The ADDR column carries the container IP (e.g. 192.168.64.2/24).
            for tok in parts:
                if tok.count(".") == 3:
                    return tok.split("/")[0], BUILDER_GUEST_PORT
    return None


def _wait_reachable(
    runtime: str,
    timeout_s: float = BUILDER_START_TIMEOUT_S,
    interval_s: float = BUILDER_POLL_INTERVAL_S,
    _sleep=time.sleep,
    _now=time.monotonic,
) -> Optional[Tuple[str, int]]:
    """Poll until the builder's sshd answers; return its (host, port) or None."""
    deadline = _now() + timeout_s
    while _now() < deadline:
        addr = reachable_address(runtime)
        if addr and reachable(addr[0], addr[1]):
            return addr
        _sleep(interval_s)
    addr = reachable_address(runtime)
    return addr if (addr and reachable(addr[0], addr[1])) else None


def stop(runtime: str, name: str = BUILDER_CONTAINER) -> None:
    """Tear down the builder container (best-effort; --rm cleans the rest)."""
    cli = "container" if runtime == "container" else runtime
    with contextlib.suppress(OSError, subprocess.SubprocessError):
        subprocess.run([cli, "rm", "-f", name], capture_output=True, timeout=15)


@contextlib.contextmanager
def builder_session(runtime: str, _run=None) -> Iterator[Optional[str]]:
    """Start the builder container, yield its ``--builders`` line, tear it down.

    Ephemeral by construction: the container is removed on exit (``--rm`` +
    explicit stop), so nothing lingers or holds RAM between builds.  Yields
    ``None`` if the builder couldn't be started/reached — the caller falls back
    (QEMU builder, or a clear error).  ``_run`` is injectable for tests.
    """
    run = _run or subprocess.run
    pubkey = ensure_keypair(_run=run)
    # Fresh start each session — a stale container from a crashed run would
    # hold the name / a dead port.
    stop(runtime)
    started = False
    try:
        run(pull_argv(runtime), capture_output=True, timeout=600)
        proc = run(
            run_argv(runtime, pubkey), capture_output=True, text=True, timeout=60
        )
        started = getattr(proc, "returncode", 1) == 0
        if not started:
            yield None
            return
        addr = _wait_reachable(runtime)
        if addr is None:
            yield None
            return
        yield builders_line(addr[0], addr[1])
    finally:
        if started:
            stop(runtime)

File: internal/containerbuilder/containerbuilder.go
// Package containerbuilder provides the on-demand container-based Linux builder for
// the macOS container runtimes. When a `packages:` build isn't cached, nix must
// offload to Linux; instead of a second hypervisor, a tiny nix+sshd builder runs
// as a container on the runtime already up, and the host drives it via nix's
// ssh-ng remote-builder protocol. The argv/URI/`builders`-line builders and the
// `container ls` ADDR parse are byte-exact contracts (the nix --builders line is
// as runbook-critical as internal/builder's); the pull/run/wait/stop lifecycle
// and the session context manager stay in the run wiring until the macos_user
// port lands.
package containerbuilder

import (
	"fmt"
	"strings"

	"github.com/mschulkind-oss/yolo-jail/internal/paths"
)

// Frozen constants (byte-identical to container_builder.py).
const (
	BuilderImage     = "ghcr.io/mschulkind-oss/yolo-jail-builder:latest"
	BuilderContainer = "yolo-linux-builder"
	BuilderSSHUser   = "root"
	BuilderGuestPort = 22
	BuilderHostPort  = 31022
)

// BuilderKeyDir is the per-workspace host-daemon key dir under GLOBAL_STORAGE.
// that reads GLOBAL_STORAGE at import).
func BuilderKeyDir() string {
	return paths.GlobalStorage() + "/linux-builder-container"
}

// BuilderKey is the private-key path; its .pub half is authorized in the
// container.
func BuilderKey() string {
	return BuilderKeyDir() + "/id_ed25519"
}

// PullArgv returns the argv to pull the builder image on the given runtime.
func PullArgv(runtime, image string) []string {
	if image == "" {
		image = BuilderImage
	}
	if runtime == "container" {
		return []string{"container", "image", "pull", image}
	}
	return []string{runtime, "pull", image}
}

// RunArgv returns the argv to start the builder container detached. podman
// publishes sshd to 127.0.0.1:<hostPort>; Apple Container has no -p (each
// container gets its own VM IP), so the publish is omitted there. Mirrors
// run_argv. Empty image/name/0 hostPort fall back to the frozen defaults.
func RunArgv(runtime, pubkey, image, name string, hostPort int) []string {
	if image == "" {
		image = BuilderImage
	}
	if name == "" {
		name = BuilderContainer
	}
	if hostPort == 0 {
		hostPort = BuilderHostPort
	}
	common := []string{
		"run", "-d", "--rm", "--name", name,
		"-e", "YOLO_BUILDER_PUBKEY=" + pubkey,
	}
	if runtime == "container" {
		argv := []string{"container"}
		argv = append(argv, common...)
		return append(argv, image)
	}
	argv := []string{runtime}
	argv = append(argv, common...)
	argv = append(argv, "-p", fmt.Sprintf("127.0.0.1:%d:%d", hostPort, BuilderGuestPort), image)
	return argv
}

// BuilderURI is the ssh-ng store/builder URI nix uses to reach the container.
func BuilderURI(host string, port int, keyPath string) string {
	if port == 0 {
		port = BuilderHostPort
	}
	if keyPath == "" {
		keyPath = BuilderKey()
	}
	return fmt.Sprintf("ssh-ng://%s@%s:%d?ssh-key=%s", BuilderSSHUser, host, port, keyPath)
}

// BuildersLine is a nix --builders spec pointing at the container. Format:
// "ssh-ng://user@host:port aarch64-linux key maxjobs". System is fixed to
// aarch64-linux (the arch a Mac needs).
// to BuilderKey() when empty; port 0 / maxJobs 0 fall back to defaults.
func BuildersLine(host string, port, maxJobs int, keyPath string) string {
	if port == 0 {
		port = BuilderHostPort
	}
	if maxJobs == 0 {
		maxJobs = 4
	}
	if keyPath == "" {
		keyPath = BuilderKey()
	}
	return fmt.Sprintf("ssh-ng://%s@%s:%d aarch64-linux %s %d", BuilderSSHUser, host, port, keyPath, maxJobs)
}

// NixSSHOpts is the NIX_SSHOPTS for talking to an ephemeral container (no
// host-key pinning — the container regenerates its key each boot). Mirrors
// nix_ssh_opts byte-for-byte.
func NixSSHOpts() string {
	return "-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
}

// ReachableAddressFromContainerLs parses `container ls` stdout for the running
// builder's VM IP:22 (Apple Container has no host port-publish).
// container branch of reachable_address: skip header, find the row whose first
// field == name, then the first token containing exactly 3 dots is the ADDR
// (e.g. "192.168.64.2/24"); strip the "/mask". Returns (host, port, true) or
// (,,false). The podman branch (always 127.0.0.1:BUILDER_HOST_PORT) is a
// constant the caller applies directly.
func ReachableAddressFromContainerLs(stdout, name string) (string, int, bool) {
	trimmed := strings.TrimSpace(stdout)
	if trimmed == "" {
		return "", 0, false
	}
	lines := strings.Split(trimmed, "\n")
	if len(lines) <= 1 {
		return "", 0, false
	}
	for _, line := range lines[1:] { // skip header
		parts := strings.Fields(line)
		if len(parts) == 0 || parts[0] != name {
			continue
		}
		for _, tok := range parts {
			if strings.Count(tok, ".") == 3 {
				ip := strings.SplitN(tok, "/", 2)[0]
				return ip, BuilderGuestPort, true
			}
		}
	}
	return "", 0, false
}
